### ISSUE TITLE
Constrain (0, max) instead of (-max, max)?

### ISSUE THREAD

**Issue by EdeMeijer** [batzner/indrnn#1]
https://github.com/batzner/indrnn/blob/551f9fe06e5c955f586a6cceccbd4d0bbacebc20/ind_rnn_cell.py#L68

Hey, I checked your implementation of the paper and noticed that instead of constraining the recurrent weight between 0 and max, where max is pow(2, 1/T), you constrain between -max and max. Since this matrix is applied element wise for every time step, wouldn't negative weights potentially result in outputs oscillating between positive and negative signs? This might explain why your version did not converge as fast as in the paper.

If this is the case, the I think the standard weight initialization might not be optimal since it's centered around 0, so half of the weights would be immediately truncated to 0. Maybe a uniform distribution between 0 and max might help initial convergence. Just a thought.

Let me know what you think. I feel like this architecture night be very promising because of it's simplicity and I'd love to see more results.

**Comment by EdeMeijer** [comment]
I tried implementing my suggestion and got much better results for the 100 length addition task. 

![compare](https://user-images.githubusercontent.com/5758565/37448338-335eb04e-2826-11e8-91a2-5ac18ca91ef6.png)

I'll send a PR your way

**Comment by batzner** [comment]
@EdeMeijer thank you for the hint! I took a look at the paper and they state that 

> the recurrent weight is constrained in the range of |u_n| in (0, pow(2,1/T))

Because they used the absolute value of u_n, I interpreted it as -pow(2,1/T) being the lower bound. Very interesting, that you got far better results with u_n in (0, pow(2,1/T).

**Comment by EdeMeijer** [comment]
I see, good point. Maybe I should run it a few more times and the faster convergence was random, but I'm sure that the initialization helped a lot though. Still, I think negative values would cause oscillation and probably wouldn't be very helpful. How I see it, these weights kind of act like the forget gates in an LSTM which are also always bounded between 0 and with a sigmoid activation.

**Comment by EdeMeijer** [comment]
In fact, I removed the constraint and let the weights initialize between [0, 1] uniformly and it converged just fine.

**Comment by batzner** [comment]
I tried uniform [0,1] recurrent inits as well and they seem to make a lot of difference. Would be nice to know the actual values used in the paper. I suggest we change the PR to keep the uniform recurrent initializer and after that, I can open a new PR that fixes the recurrent weight clipping?

I just noticed, that simply clipping to `[-recurrent_max, recurrent_max]` is not what the paper suggests. They suggest that the weights either lie in` [-recurrent_max, -recurrent_min]` or in `[recurrent_min, recurrent_max]`. Although, for the ReLU default of (0, pow(2, 1/T)), that doesn't make a difference.
